### ISSUE TITLE
Dataslayer causing recursive JS errors, bricking the tab.

### ISSUE THREAD

**Issue by cthae** [sean-adams/dataslayer#115]
**Please include, if available:**
- the URL of the page with an error: https://shop.lululemon.com/
- any errors in the console for the page: It's really easy to reproduce, but pain to copy. Check it out.
- your Chrome version and dataslayer version: latest chrome and dataslayer.

I would be interested in knowing the underlying issue of this all. Also how do we stop a recursive issue like that from happening, just to debug it?

**Comment by sean-adams** [comment]
Hi @cthae sorry for the delayed response. I have had way less time to work on dataslayer this year. Confirming this (very annoying, sorry!) bug which seems to be an issue with how dataslayer monitors Adobe Launch rules. I will do some more investigation ASAP and hopefully resolve this.

**Comment by sean-adams** [comment]
Hi @cthae I've looked into this a bit more, this seems to be a result of some suboptimal coding on both of our parts :slightly_smiling_face: 

In order to display the values of Launch data elements in the dataslayer panel, we have to query them from Launch using the `_satellite.getVar()` method. This is accomplished with a `for` loop over all of the data elements we get from Launch (code significantly condensed):

```js
    for (var i = 0; i < elementNames.length; i++) {
      let cleanValue = window._satellite.getVar(elementNames[i]);
      ...
      launchElements[elementNames[i]] = cleanValue;
    }
```

Through isolating this section of code and running it on smaller subsets of `elementNames` I was eventually able to identify that the `cartOOSSKUArray` Launch data element appears to be the one that triggers this problem.

Note in the above code the ` for (var i = 0` and how often dataslayer uses the `i` value as it iterates through the list of elements.

Now, a section of the code for `cartOOSSKUArray`:
```js
                for (i = 0; i < t.length; i++)
                    r.push(t[i].getAttribute("data-skuid"));
```

Unfortunately it looks like the use of `i` here without redeclaring it is causing the `i` being used by dataslayer's loop to reset to 0 each time this rule is executed, causing the infinite loop.

You should be able to fix this error in your code by changing `for (i = 0; i < t.length; i++)` to `for (var i = 0; i < t.length; i++)`.

This is a super annoying thing about declaring variables with `var` and why block-scoped variables declared with `const` and `let` are a big improvement. Unfortunately a significant chunk of the dataslayer code was written when I was a much more junior developer :slightly_smiling_face: I will fix this specific issue immediately and look to strip out as much code as possible throughout dataslayer that uses potentially problematic variable scoping.

**Comment by sean-adams** [comment]
Actually, there's something more going on here beyond that. Will update again.

**Comment by sean-adams** [comment]
OK, so while the above item was causing some issues, the call stack issue you originally flagged was another unfortunate interaction between dataslayer and your implementation.

As I noted above:

> In order to display the values of Launch data elements in the dataslayer panel, we have to query them from Launch using the _satellite.getVar() method.

It seems like your implementation fires some additional Launch rules when certain of these elements fails. This is causing poor performance as dataslayer queries these values:
- on a timer
- when a Launch rule is completed firing
to ensure it can provide the most up-to-date data. But in the case of this implementation, this only triggers more errors and data.

To mitigate this:
- next version of dataslayer will allow you to disable updating these elements when a Launch rule is fired (_"don't update Launch elements on rules"_)
- next version will also allow you to manually refresh these elements
- I recommend you disable the timer (_"Update interval (seconds)"_) for this implementation

**Comment by sean-adams** [comment]
@cthae the mitigations noted above will be available in version 1.3.8 once that is approved by the Chrome web store team.
